web/user: refactor LibraryPage for testing, add CTA

kensternberg-authentik opened the pull request:
## Details

-   Resolves #5171

## Changes

web: revise LibraryPage, add CTA

### Summary

TL;DR:

- Separated LibraryPage into a bunch of different, independent parts, none of which require Authentik running to be testable or viewable.
- This made adding the “Add an Application” CTA easier.
- This sets the stage for unit and view testing of the UI

### Description

#### Refactored

This commit revises the LibraryPage, devolving it into a couple of independent components that have to asynchronous dependencies, with a single asynchronous master:

- LibraryPage: Loads the UIConfig, UserConfig, and CoreApi, and once those are loaded, launches the LibraryPageImpl.
- LibraryPageImpl: the ListView of applications available, and updates the ListView according to search criteria it receives via an event listener.
  - LibraryPageImpl.css: The stylesheet. Put here because it’s visual clutter.
  - LibraryPageImpl.utils: defines static functions used to filter the view. Here because, again, it would otherwise be visual clutter of the LibraryPageImpl.
- ApplicationEmptyState: Shows the “You have no applications” and, if the user is a superuser, the “Add an application” button.
- ApplicationSearch: Contains the Fuse implementation and, as the search result is updated, sends the selected and filtered app list to the LibraryPage via an event. Also controls the “Choose an application by pressing Enter” event.
- ApplicationList: Displays the list of applications.

All of these components are *responsive* to changes in the Apps collection via the LibraryPage itself, but none of them invoke the Apps collection, UIConfig, and CoreApi directly, so it should be possible to create Storybook implementations that view the LibraryPageImpl itself without having to have an instance of Authentik running.

#### Extension and Bugfix

If the user is a superuser, the “You have no applications” panel now shows the “Add an Application” button and a link to the documentation on how to add an application.

### New Features

-   Adds a "Add an Application" to the LibraryView if there are no applications and the user is an administrator.

### Breaking Changes

No breaking changes.

## Checklist

-   [X] Local tests pass (`ak test authentik/`)
-   [X] The code has been formatted (`make lint-fix`)

If an API change has been made

-   No API  Changes were made.

If changes to the frontend have been made

-   [X] The code has been formatted (`make web`)
-   [X] The translation files have been updated (`make i18n-extract`)

No documentation changes were made.

kensternberg-authentik committed:
5171: Fixed README to comply with Prettier rules.

I'm pretty impressed that that worked.  Good on Jens for having that
in the prettier rules.

kensternberg-authentik committed:
web: revised package.json

Removed the migration and web/README.md file.  The former should not have
been included; the latter is currently unprofessional in tone.

kensternberg-authentik committed:
web: revise LibraryPage, add CTA

TL;DR:

- Separated LibraryPage into a bunch of different, independent parts, none of which require
Authentik running to be testable or viewable.
- This made adding the "Add an Application" CTA easier.
- This sets the stage for unit and view testing of the UI

This commit revises the LibraryPage, devolving it into a couple of independent components that have
to asynchronous dependencies, with a single asynchronous master:

- LibraryPage: Loads the UIConfig, UserConfig, and CoreApi, and once those are loaded, launches the
  LibraryPageImpl.
- LibraryPageImpl: the ListView of applications available, and updates the ListView according to
  search criteria it receives via an event listener.
  - LibraryPageImpl.css: The stylesheet. Put here because it's visual clutter.
  - LibraryPageImpl.utils: defines static functions used to filter the view. Here because, again, it
    would otherwise be visual clutter of the LibraryPageImpl.
- ApplicationEmptyState: Shows the "You have no applications" and, if the user is a superuser, the
  "Add an application" button.
- ApplicationSearch: Contains the Fuse implementation and, as the search result is updated, sends
  the selected and filtered app list to the LibraryPage via an event.  Also controls the "Choose an
  application by pressing Enter" event.
- ApplicationList: Displays the list of applications.

All of these components are _responsive_ to changes in the Apps collection via the LibraryPage
itself, but none of them invoke the Apps collection, UIConfig, and CoreApi directly, so it should be
possible to create Storybook implementations that view the LibraryPageImpl itself without having to
have an instance of Authentik running.

If the user is a superuser, the "You have no applications" panel now shows the "Add an Application"
button and a link to the documentation on how to add an application.

kensternberg-authentik committed:
web: lint and prettier updates

\#\# Details

-   Resolves #5171

\#\# Changes

This just updates the prettier and eslint passes.

kensternberg-authentik committed:
\#\# Details

-   Resolves #5171

\#\# Changes

Removed unused declarations.

BeryJu reviewed (commented):
A bunch of nitpicks and a bunch of opinions/questions, already looking very good and should be a lot easier to test

kensternberg-authentik committed:
\#\# Details

-   web: refactor LibraryPage, resolves #5171

\#\# Changes

Some changes found in code review, including an embarassing failure
to both remove the old internal accessor and propagate the new
one for "isAdmin".

A pattern is emerging that a LitComponent class should consist of:

- styles
- properties
- states
- queries
- other object fields
- constructor()
- connectedCallBack()
- disconnectedCallBack()
- event listeners
- callback helpers
- render helpers
- render()

... in that order.

BeryJu committed:
actually remove LibraryPage that got re-added in the rebase

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

BeryJu committed:
fix router import

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

BeryJu committed:
use pf-c-button for CTA

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

BeryJu committed:
fix different alignment compared to old version

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

BeryJu committed:
use docLink() for documentation link

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

BeryJu committed:
also open docs in new tab

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

BeryJu commented on web/src/user/LibraryPage/types.ts:
```diff
@@ -0,0 +1,12 @@
+export type AppGroupEntry = [string, Application[]];
+export type AppGroupList = AppGroupEntry[];
+
+export type PageUIConfig = {
```
Yeah thats a fair point, allthough maybe it would be better to go the other way around and compose `UIConfig` of smaller interfaces, and we don't have to add things in two places, but that can be a later PR if we want to do that

tanberry commented:
Thanks @Ken for adding  CTA here! One request: can you change the wording about the Docs to "Refer to documentation"?

kensternberg-authentik committed:
web: minor language changes

As requested by @Tana.
